lib/services/suppressions-service.js

```diff
@@ -0,0 +1,270 @@
+
+            for (const ruleId in violationsByRule) {
+                if (!unused[relativeFilePath][ruleId]) {
+                    continue;
+                }
+
+                const suppressionsCount = unused[relativeFilePath][ruleId].count;
+                const violationsCount = violationsByRule[ruleId].count;
+
+                // Suppress messages if the number of violations is less or equal to the suppressions count
+                if (violationsCount <= suppressionsCount) {
+                    SuppressionsService.suppressMessagesByRule(result, ruleId);
+                }
+
+                // Update the count to match the new number of violations, otherwise remove the rule entirely
+                if (violationsCount < suppressionsCount) {
+                    unused[relativeFilePath][ruleId].count = suppressionsCount - violationsCount;
+                } else {
+                    delete unused[relativeFilePath][ruleId];
+                }
+            }
+
+            // Cleanup files with no rules
+            if (Object.keys(unused[relativeFilePath]).length === 0) {
+                delete unused[relativeFilePath];
+            }
+        }
+
+        // Remove entries from unused that were not included in the results
+        for (const file in unused) {
+            if (!filtered.some(result => this.getRelativeFilePath(result.filePath) === file)) {
+                delete unused[file];
+            }
+        }
+
+        return {
+            results: filtered,
+            unused
+        };
+    }
+
+    /**
+     * Loads the suppressions file.
+     * @throws {Error} If the suppressions file cannot be parsed.
+     * @returns {Promise<SuppressedViolations>} The suppressions.
+     */
+    async load() {
+        try {
+            const data = await fs.promises.readFile(this.filePath, "utf8");
+
+            return JSON.parse(data);
+        } catch (err) {
+            if (err.code === "ENOENT") {
+                return {};
+            }
+            throw new Error(`Failed to parse suppressions file at ${this.filePath}`);
+        }
+    }
+
+    /**
+     * Updates the suppressions file.
+     * @param {SuppressedViolations} suppressions The suppressions to save.
+     * @returns {Promise<void>}
+     * @private
+     */
+    save(suppressions) {
+        return fs.promises.writeFile(
+            this.filePath,
+            JSON.stringify(suppressions, null, 2)
+        );
+    }
+
+    /**
+     * Counts the violations by rule, ignoring warnings.
+     * @param {LintMessage[]} messages The messages to count.
+     * @returns {Record<string, number>} The number of violations by rule.
+     */
+    static countViolationsByRule(messages) {
+        return messages.reduce((totals, message) => {
+            if (message.severity === 2) {
+                totals[message.ruleId] ??= { count: 0 };
+                totals[message.ruleId].count++;
+            }
```

Comment:
I think this should skip messages that have `ruleId === null`. Those are, for example, messages for unused disable directives.